#162948842 Have CRUD on Articles Resource

KMaina opened the pull request:
#### What does this PR do?
Have CRUD on Articles resource
#### Description of Task to be completed?

`POST api/articles/`

`GET api/articles/`

`GET api/articles/:slug`

`PUT api/articles/:slug`

`PUT api/articles/delete/:slug`

#### How should this be manually tested?

- Clone the repo locally
- cd ah-django-unchained
- git checkkout ft-crud-on-articles-162948842
- get started with the app as from the README section
- to add an article ensure you obtain a token from logging in or signing up if not signed up
- Using the `POST api/articles/` endpoint, have the following JSON object. Note that image uploads are optional.
```
{
    "title": "My story",
    "description": "I am very happy",
    "body": "A story about me"
}
```
<img width="1136" alt="post article" src="https://user-images.githubusercontent.com/31338414/51385132-66958a80-1b2f-11e9-8f1e-31cbd66e4850.png">

- Using the `GET api/articles/`, fetch all articles. Only those that have `is_displayed=True` will be displayed as we assume the others have been deleted.
<img width="1034" alt="get all articles" src="https://user-images.githubusercontent.com/31338414/51385308-e91e4a00-1b2f-11e9-8bf6-303480c1462f.png">

- Using the `GET api/articles/:slug` endpoint we are able to fetch a single article depending on the name of the slug
<img width="1031" alt="get single article" src="https://user-images.githubusercontent.com/31338414/51385398-37cbe400-1b30-11e9-8108-f766dbf4c6ce.png">

- Using the `PUT api/articles/:slug` we can edit an article. Use the JSON body below
```
{
    "title": "My story",
    "description": "I am very happy",
    "body": "A story about me"
}
```

- Using the `PUT api/articles/delete/:slug`, we can delete an article by setting its `is_displayed` field to `False`
```
{
    "is_displayed": False
}
```

#### What are the relevant pivotal tracker stories?
[#162948842](https://www.pivotaltracker.com/story/show/162948842)

KMaina committed:
feat(CRUD Articles): add articles app to settings

- ensure the app is registered in the app

[Delivers #162948842]

KMaina committed:
feat(CRUD Articles): add articles app

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): create models for the articles

- allows for created articles to be saved in a database

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add serializers for the Article model

- serialize data from the Article model

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): fix cloudinary imports

- import the fields from cloudinary

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add the database migration file

- effect the changes we made to the models

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): edit fields in the Article model

- allow for images to be optionally uploaded
- generate unique slugs

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add migrations made to the database

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add taggit in installed apps

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add serializers for the Article model

- Have a serializer for adding and retrieving posts
- Have a serializer for updating and retrieving a single post
- Have a serializer for deleting a post

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add tests for all article endpoints

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add routes for the article endpoints

- add a route for fetching and posting endpoints
- add routes for deleting articles
- add routes for updating endpoints

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add the route for the articles app

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add views for the articles app

- add a view to post and fetch all articles
- add a view to update an article
- add a view to delete an article

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add packages in requirements.txt

- Add Pillow and Taggit

[Deliver #162948842]

Georgeygigz commented:
Kindly rebase your work in order to resolve the conflicts

Georgeygigz commented on authors/apps/articles/test/test_create_article.py:
```diff
@@ -0,0 +1,130 @@
+from django.urls import reverse
+from rest_framework.views import status
+from rest_framework.test import APITestCase, APIClient
+# from ..models import User
```
Kindly remove the commented imports

KMaina replied:
Noted

Georgeygigz commented on authors/apps/articles/tests.py:
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
```
Are you using this file in your implementation?

KMaina replied:
I am not using that file. Let me delete it.

ignatiusukwuoma replied:
File not in use

Georgeygigz commented on authors/apps/articles/test/test_create_article.py:
```diff
@@ -0,0 +1,130 @@
+
+    def test_fetch_all_articles(self):
+        """Tests to fetch all articles"""
+        register = self.client.post(self.signup_url,
```
You can create a method that returns a token then call it to avoid repetition

KMaina replied:
Noted

Georgeygigz commented on authors/apps/articles/test/test_create_article.py:
```diff
@@ -0,0 +1,130 @@
+        response = self.client.post(self.listcreate,
+                                    self.create_article_data, format='json',
+                                    HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertEqual(register.status_code, status.HTTP_201_CREATED)
```
Then I think you are supposed to assert response instead of register.

d-kahara replied:
> Then I think you are supposed to assert response instead of register.

The response is being saved as a variable called `register`. Renaming it would perhaps avoid confusion

d-kahara commented on authors/apps/articles/serializers.py:
```diff
@@ -0,0 +1,63 @@
+                  'modified',
+                  'images',
+                  'author']
+        read_only_fields = ['created',
```
Consider adding an `is_published` field to allow saving of drafts

KMaina replied:
done

KMaina committed:
feat(CRUD Articles): add migration file for model changes

- Fix typo in is_displayed field
- add is_published field

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add migration file for model changes

- Fix typo in is_displayed field
- add is_published field
- remove the save method that added slugs upon saving

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add changes to the serializer file

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): fix typo in field name

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): Refactor tests written for CRUD on articles

- Fix typo
- Remove redundant code blocks
- Add more tests for edge cases

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): refactor code in the views.py file

- add functionality to generate unique slugs

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add postgres version in .travis.yml file

[Deliver #162948842]

ignatiusukwuoma commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,19 @@
+    This is the Article model that is used to handle CRUD on articles
+    """
+    title = models.CharField(max_length=200, blank=False)
+    body = models.TextField(blank=False, unique=True)
```
Why unique=True?

KMaina replied:
I wanted each article to be unique.

ignatiusukwuoma replied:
Doesn't have to be. I don't see any need

ignatiusukwuoma commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,19 @@
+    title = models.CharField(max_length=200, blank=False)
+    body = models.TextField(blank=False, unique=True)
+    description = models.CharField(max_length=50, blank=False)
+    images = CloudinaryField(blank=True, default='', null=True)
```
What does empty `default=''` do?

KMaina replied:
Let me remove it

ignatiusukwuoma commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,19 @@
+    created = models.DateTimeField(auto_now_add=True)
+    modified = models.DateTimeField(auto_now=True)
+    slug = models.SlugField(max_length=50, blank=False, unique=True)
+    is_displayed = models.BooleanField(default=True)
```
`is_deleted` passes the message better. In that case, default will be False

ignatiusukwuoma commented on authors/settings.py:
```diff
@@ -40,10 +41,12 @@
     'django_extensions',
     'rest_framework',
-
+    'taggit',
```
If this is for tagging articles, remove it since you're not implementing that in this task

ignatiusukwuoma commented on authors/apps/articles/test/test_create_article.py:
```diff
@@ -0,0 +1,181 @@
+import json
```
Directory name should be `tests` since you'll have more tests. The tests in this file are not only for creating articles, yet your file_name and class name says it's only about creating articles.

ignatiusukwuoma commented:
Pls run your files though pycodestyle.

KMaina committed:
feat(CRUD Articles): ensure code meets PEP8 standards

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add migration for model changes

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): add changes in model

- change name of deleted items to is_deleted
- ensure code meets PEP8 standards

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): ensure code meets PEP8 standards

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): ensure code meets PEP8 standards

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): ensure code meets PEP8 standards

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): ensure code meets PEP8 style

[Deliver #162948842]

MaggieChege commented:
In your views, in your get endpoint add a Response if there are no articles.

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -0,0 +1,183 @@
+        self.assertIn("This is my story", response.data[0]['body'])
+        self.assertIn("This begins now", response.data[1]['body'])
+        self.assertIn("Here is my story", response.data[0]['description'])
+        self.assertIn("Here is my story", response.data[1]['description'])
```
It's difficult to understand where these strings `"Here is my story"`, etc are coming from. Please use the string directly from the object, like `self.create_article_data.description`

KMaina replied:
done

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -0,0 +1,183 @@
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertIn("A new story", response.data['title'])
+        self.assertIn("This is my story", response.data['body'])
+        self.assertIn("Here is my story", response.data['description'])
```
ditto

KMaina replied:
done

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -0,0 +1,57 @@
+class ArticleAPIView(ListCreateAPIView):
+    """Creates articles and retrieves all articles"""
+    permission_classes = (IsAuthenticatedOrReadOnly,)
+    # Only fetch those articles whose 'is_deleted' field is True
```
where 'is_deleted' field is True? or False

KMaina replied:
done

KMaina committed:
feat(CRUD Articles): improve readability of tests

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): change bool value in comment

[Deliver #162948842]

ignatiusukwuoma commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,20 @@
+    """
+    title = models.CharField(max_length=200, blank=False)
+    body = models.TextField(blank=False)
+    description = models.CharField(max_length=50, blank=False)
```
I think an article should be able to be saved without the title, body or description. The can be left blank. Remember that when the user starts typing, on the frontend we will make requests to save the article.

When it’s about to be published, then we can require that title and body is present, but not necessarily the description

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -0,0 +1,198 @@
+                      response.data[1]['body'])
+        self.assertIn(self.create_article_data['description'],
+                      response.data[0]['description'])
+        self.assertIn(self.create_article_data2['description'],
```
Shouldn't these `assertIn` be `assertEquals`? I expect that the exact content I sent in will be the response.

KMaina replied:
Done

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -0,0 +1,198 @@
+                      response.data['title'])
+        self.assertIn(self.create_article_data['body'],
+                      response.data['body'])
+        self.assertIn(self.create_article_data['description'],
```
`assertIn` should be used when the first argument is a subset of the second argument

KMaina replied:
Done

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -0,0 +1,198 @@
+                      response.data['title'])
+        self.assertIn(self.create_article_data['body'],
+                      response.data['body'])
+        self.assertIn(self.create_article_data['description'],
```
ditto

KMaina replied:
Done

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -0,0 +1,58 @@
+
+from django.template.defaultfilters import slugify
+from rest_framework import status
+from django.core import serializers
```
Unused import

KMaina replied:
Removed

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -0,0 +1,58 @@
+                                     RetrieveUpdateAPIView, UpdateAPIView)
+from rest_framework.permissions import (IsAuthenticated,
+                                        IsAuthenticatedOrReadOnly)
+from rest_framework.renderers import JSONRenderer
```
Unused import

KMaina replied:
Removed

ignatiusukwuoma commented on authors/apps/articles/views.py:
```diff
@@ -0,0 +1,58 @@
+from rest_framework.renderers import JSONRenderer
+
+from authors.apps.core.permissions import IsAuthorOrReadOnly
+from . import serializers
```
Unused import

KMaina replied:
Removed

KMaina committed:
feat(CRUD Articles): change assertion methods in tests

- change from assertIn to assertEqual

[Deliver #162948842]

KMaina committed:
feat(CRUD Articles): fix failing test

[Deliver #162948842]

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -0,0 +1,199 @@
+        response = self.client.get(reverse('articles:articles-retrieveupdate',
+                                   kwargs={'slug': 'a-new-story'}),
+                                   format='json')
+        print(response.data)
```
Remove print statement

KMaina replied:
Done

ignatiusukwuoma commented on authors/apps/articles/tests/test_create_article.py:
```diff
@@ -0,0 +1,199 @@
+                                   self.delete_article,
+                                   format='json',
+                                   HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
assert that the `is_deleted` field of the article is now true

KMaina replied:
Done

KMaina committed:
feat(CRUD Articles): add test case for testing articles

[Deliver #162948842]
